File: Atividade 2/serialization.py
"""
Este módulo contém os métodos de serialização e desserialização do servidor.
"""

from struct import pack, unpack
from person import *

FORMAT = 'utf-8'
SIZE_OF_INT = 4
SIZE_OF_LONG = 8


def deserialize_int(connection):
    return unpack('!I', connection.recv(SIZE_OF_INT))[0]

def deserialize_long(connection):
    return unpack('!Q', connection.recv(SIZE_OF_LONG))[0]

def deserialize_string(connection):
	string_lenght = deserialize_int(connection)
	return connection.recv(string_lenght).decode(FORMAT)

def deserialize_person(connection):
    for _ in range(0, 3):
        person_field = deserialize_string(connection)
        match person_field:
            case 'nome':
                name = deserialize_string(connection)
            case 'endereco':
                for _ in range(0, 3):
                    address_field = deserialize_string(connection)
                    match address_field:
                        case 'rua':
                            street = deserialize_string(connection)
                        case 'bairro':
                            neighborhood = deserialize_string(connection)
                        case 'numero':
                            number = deserialize_long(connection)
                        case _:
                            pass
            case 'dados_bancarios':
                for _ in range(0, 3):
                    bank_details_field = deserialize_string(connection)
                    match bank_details_field:
                        case 'banco':
                            bank = deserialize_string(connection)
                        case 'agencia':
                            agency = deserialize_string(connection)
                        case 'conta':
                            account = deserialize_string(connection)
                        case _:
                            pass
            case _:
                pass
    return Person(name, Address(street, neighborhood, number), BankDetails(bank, agency, account))

def serialize_int(connection, number):
    connection.send(pack('!I', number))

def serialize_long(connection, number):
    connection.send(pack('!Q', number))

def serialize_string(connection, string):
    serialize_int(connection, len(string))
    connection.send(string.encode(FORMAT))

def serialize_person(connection, person):
    serialize_string(connection, "nome")
    serialize_string(connection, person.name)

    serialize_string(connection, "endereco")
    serialize_string(connection, "rua")
    serialize_string(connection, person.address.street)
    serialize_string(connection, "bairro")
    serialize_string(connection, person.address.neighborhood)
    serialize_string(connection, "numero")
    serialize_long(connection, person.address.number)

    serialize_string(connection, "dados_bancarios")
    serialize_string(connection, "banco")
    serialize_string(connection, person.bank_details.bank)
    serialize_string(connection, "agencia")
    serialize_string(connection, person.bank_details.agency)
    serialize_string(connection, "conta")
    serialize_string(connection, person.bank_details.account)

File: Atividade 2/serialization.c
#include "serialization.h"
#include <stdlib.h>
#include <string.h>

#define MAX_STR_LEN 256

void serialize_int(int client_socket, uint32_t number) {
	uint32_t number_network_order = htonl(number);
	send(client_socket, &number_network_order, sizeof(number_network_order), 0);
}

void serialize_long(int client_socket, uint64_t number) {
	uint32_t high_part = htonl((uint32_t) (number >> 32));
	uint32_t low_part = htonl((uint32_t) (number & 0xFFFFFFFF));
	send(client_socket, &high_part, sizeof(high_part), 0);
	send(client_socket, &low_part, sizeof(low_part), 0);
}

void serialize_string(int client_socket, const char *string) {;
	uint32_t length = htonl(strlen(string));
	send(client_socket, &length, sizeof(length), 0);
	send(client_socket, string, strlen(string), 0);
}

void serialize_person(int client_socket, Pessoa pessoa) {
	serialize_string(client_socket, "nome");
	serialize_string(client_socket, pessoa.nome);

	serialize_string(client_socket, "endereco");
	serialize_string(client_socket, "rua");
	serialize_string(client_socket, pessoa.endereco.rua);
	serialize_string(client_socket, "bairro");
	serialize_string(client_socket, pessoa.endereco.bairro);
	serialize_string(client_socket, "numero");
	serialize_long(client_socket, pessoa.endereco.numero);

	serialize_string(client_socket, "dados_bancarios");
	serialize_string(client_socket, "banco");
	serialize_string(client_socket, pessoa.dados_bancarios->banco);
	serialize_string(client_socket, "agencia");
	serialize_string(client_socket, pessoa.dados_bancarios->agencia);
	serialize_string(client_socket, "conta");
	serialize_string(client_socket, pessoa.dados_bancarios->conta);
}

int deserialize_int(int client_socket) {
	uint32_t number_network_order;
	recv(client_socket, &number_network_order, sizeof(number_network_order), 0);
	int number = ntohl(number_network_order);
	return number;
}

long deserialize_long(int client_socket) {
    uint32_t high_part, low_part;
    recv(client_socket, &high_part, sizeof(high_part), 0);
    recv(client_socket, &low_part, sizeof(low_part), 0);
    high_part = ntohl(high_part);
    low_part = ntohl(low_part);
    long number = ((long) high_part << 32) | low_part;  
    return number;
}

char *deserialize_string(int client_socket) {
	uint32_t length_network_order;
	recv(client_socket, &length_network_order, sizeof(length_network_order), 0);
	uint32_t length = ntohl(length_network_order);
	char *string = (char*) malloc(length + 1);
	recv(client_socket, string, length, 0);
	string[length] = '\0';
	return string;
}

Pessoa deserialize_person(int client_socket) {
    int i, j;
    char *person_field, *address_field, *bank_details_field;
    Pessoa pessoa = create_person();

    for (i = 0; i < 3; i++) {
        person_field = deserialize_string(client_socket);

        if (!strcmp(person_field, "nome")) {
            pessoa.nome = deserialize_string(client_socket);
        }

        else if (!strcmp(person_field, "endereco")) {
            for (j = 0; j < 3; j++) {
                address_field = deserialize_string(client_socket);
                if (!strcmp(address_field, "rua"))
                    pessoa.endereco.rua = deserialize_string(client_socket);
                else if (!strcmp(address_field, "bairro"))
                    pessoa.endereco.bairro = deserialize_string(client_socket);
                else if (!strcmp(address_field, "numero"))
                    pessoa.endereco.numero = deserialize_long(client_socket);
            }
        }

        else if (!strcmp(person_field, "dados_bancarios")) {
            for (j = 0; j < 3; j++) {
                bank_details_field = deserialize_string(client_socket);
                if (!strcmp(bank_details_field, "banco"))
                    pessoa.dados_bancarios->banco = deserialize_string(client_socket);
                else if (!strcmp(bank_details_field, "agencia"))
                    pessoa.dados_bancarios->agencia = deserialize_string(client_socket);
                else if (!strcmp(bank_details_field, "conta"))
                    pessoa.dados_bancarios->conta = deserialize_string(client_socket);
            }
        }
    }

    return pessoa;
}
